File: packs/dsa-two-pointers-ts/challenges/04b-valid-palindrome-reverse.test.ts
import { join } from 'node:path';
import { writeChallengeToTmp, importModule } from '../../test-helpers.js';

const CHALLENGE = join(import.meta.dirname, '04b-valid-palindrome-reverse.json');

describe('Valid Palindrome (Reverse)', () => {
  let isPalindrome: (s: string) => boolean;
  let cleanup: () => void;

  beforeAll(async () => {
    const tmp = writeChallengeToTmp(CHALLENGE);
    cleanup = tmp.cleanup;
    const mod = await importModule<{ isPalindrome: typeof isPalindrome }>(
      join(tmp.tmpDir, 'solution.ts')
    );
    isPalindrome = mod.isPalindrome;
  });

  afterAll(() => cleanup());

  it('sentence palindrome', () => {
    expect(isPalindrome('A man, a plan, a canal: Panama')).toBe(true);
  });

  it('not a palindrome', () => {
    expect(isPalindrome('race a car')).toBe(false);
  });

  it('whitespace only', () => {
    expect(isPalindrome(' ')).toBe(true);
  });

  it('mixed case with digits', () => {
    expect(isPalindrome('0P')).toBe(false);
  });

  it('empty string', () => {
    expect(isPalindrome('')).toBe(true);
  });

  it('single character', () => {
    expect(isPalindrome('a')).toBe(true);
  });

  it('punctuation only', () => {
    expect(isPalindrome('.,!')).toBe(true);
  });
});

File: packs/dsa-two-pointers-py/challenges/04b-valid-palindrome-reverse.test.py
from solution import is_palindrome


def test_sentence_palindrome():
    assert is_palindrome("A man, a plan, a canal: Panama") is True


def test_not_palindrome():
    assert is_palindrome("race a car") is False


def test_empty_like_string():
    assert is_palindrome(" ") is True


def test_mixed_case_digits():
    assert is_palindrome("0P") is False


def test_empty_string():
    assert is_palindrome("") is True


def test_single_char():
    assert is_palindrome("a") is True


def test_punctuation_only():
    assert is_palindrome(".,!") is True
